File: Insert_Node_in_a_Binary_Search_Tree.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param root: The root of the binary search tree.
     * @param node: insert this node into the binary search tree
     * @return: The root of the new binary search tree.
     */
    TreeNode* insertNode(TreeNode* root, TreeNode* node) {
        // write your code here
        //recursive
        // if (root == nullptr){
        //     return node;
        // }
        // if (node->val > root->val){
        //     root->right = insertNode(root->right, node);
        // }else if (node->val < root->val){
        //     root->left = insertNode(root->left, node);
        // }
        // return root;
        if (!root) return node;
        TreeNode *parent = root;
        while (true){
            if (parent->val > node->val){
                if (!parent->left){
                    parent->left = node;
                    break;
                }else {
                    parent = parent->left;
                }
            }else if (parent->val < node->val){
                if (!parent->right){
                    parent->right = node;
                    break;
                }else {
                    parent = parent->right;
                }
            }else {
                break;
            }
        }
        return root;
    }
};

File: Insert Node in a Binary Search Tree.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    """
    @param root: The root of the binary search tree.
    @param node: insert this node into the binary search tree.
    @return: The root of the new binary search tree.
    """
    def insertNode(self, root, node):
        # write your code here
        
        p = root
        while p:
            pre = p
            if p.val > node.val:
                if not p.left:
                    p.left = node
                    return root
                else:
                    p = p.left
            else:
                if not p.right:
                    p.right = node
                    return root
                else:    
                    p = p.right
        return node